controller/readonly.go

```diff
@@ -60,16 +60,40 @@ func (ro *ReadOnly) Init(ctx context.Context) error {
 func (ro *ReadOnly) Run(ctx context.Context) error {
 	ro.logger.Info("inspecting all tasks")
 
-	// Stop watchiing dependencies after inspect-mode completes
-	ctxDep, cancel := context.WithCancel(ctx)
-	defer cancel()
+	// Stop watching dependencies after inspecting tasks ends
+	ctxWatch, cancelWatch := context.WithCancel(ctx)
+
+	// Stop inspecting tasks early if WatchDep errors
+	ctxInspect, cancelInspect := context.WithCancel(ctx)
+
+	exitBufLen := 2 // watchDep & once-ing tasks
+	exitCh := make(chan error, exitBufLen)
 
 	// start watching dependencies in order to render templates to plan tasks
-	// once
-	go ro.tasksManager.WatchDep(ctxDep)
+	go func() {
+		exitCh <- ro.tasksManager.WatchDep(ctxWatch)
+		cancelInspect()
+	}()
 
 	// always inspect consecutively to keep inspect logs in order
-	return ro.inspectConsecutive(ctx)
+	go func() {
+		exitCh <- ro.inspectConsecutive(ctxInspect)
+		cancelWatch()
+	}()
+
+	counter := 0
+	for {
+		err := <-exitCh
+		counter++
+		if err != context.Canceled {
```

Comment:
Will this be an issue if `err` is nil in the happy path case?

Reply:
If the err is nil, the counter will still increment. Once both WatchDep and inspectOnce finish, they'll both send a nil to the exitCh. Then Run() should exit at line 94 below. `Test_ReadOnly_Run` also should capture the happy test path and looks to be still passing (I did break the equivalent `Test_ReadWrite_Once_WatchDep_errors_Terraform` at one point while developing :D ). Let me know if you have concerns

Reply:
This is the scenario I was thinking of:

1. inspectConsecutive finishes, returns nil on exitCh and cancels watch
2. This loop gets the nil on the exitCh, increments counter to 1, but since nil isn't equal to `context.Canceled`, it returns nil on line 90.
3. Watch is stopped because of the cancel, but the function has already returned and didn't necessarily wait for the watch to finish cancelling.

But I also need to learn more about contexts and cancelling contexts so I might be way off base here!

Reply:
Ah, I get what you're saying. Yes, I agree. You're absolutely right. To confirm, I ran Test_ReadOnly_run with coverage and saw that it exited early and did not wait for WatchDep:

<img width="370" alt="Screen Shot 2022-04-20 at 1 07 57 PM" src="https://user-images.githubusercontent.com/6819378/164292073-6b2eca17-9ae9-4956-8cb5-4678776e08b8.png">

I added a commit to fix this for rw.Once and ro.Run (and added a cautionary check in rw.Run - from a previous change). The coverage now looks something like:
<img width="732" alt="Screen Shot 2022-04-20 at 1 52 49 PM" src="https://user-images.githubusercontent.com/6819378/164292497-a7a15a7d-8747-4ed7-a8a0-84e09d09c11b.png">

Thank you for catching this! Let me know if you have any other feedback